### DISCUSSION TITLE
tycho 2.7.1: jar is tried to get from mvn central but is (and should) just be get from the target file (p2 site)

### DISCUSSION THREAD

**Discussion by jcompagner** [eclipse-tycho/tycho#895] [General]
below a -X output

this is very weird i have no idea how tycho knows this or gets that info from
you see that it tries to resolve "com.servoy:javax.websocket"

Orignally that javax.websocket is created yes under that group. To generate our p2 site for 3rd party libs 
so it is there:

https://developer.servoy.com/p2_repository/ship_plugins/2022.06/plugins/javax.websocket_1.1.0.jar

that 2022.06 is a dir that is a p2 site that is generated from a category.xml file:
https://github.com/Servoy/servoy-eclipse/blob/master/shipplugins/category.xml#L32
and the pom: 
https://github.com/Servoy/servoy-eclipse/blob/master/shipplugins/pom.xml#L197

thats the only place where i mention the grouping..

After that the below project is build with the target file that points to that p2 site:

https://github.com/Servoy/servoy-eclipse/blob/master/launch_targets/com.servoy.eclipse.target.target#L15

but suddenly when i build my project that has that target:
https://github.com/Servoy/sablo/blob/master/sablo/pom.xml#L188

as its "classpath"

That pom has some mvn depedencies (mostly for testing) but it does not mention that javax.websockets at all.
But suddenly it seems to know that it is com.servoy grouping and it want to download it

[DEBUG]    com.servoy:javax.websocket:jar:1.1.0:compile

but that should also just be like the others so i guess something like this?

[DEBUG]    p2.eclipse.plugin:javax.websocket:jar:1.1.0:system

because it should just be gotten from the p2 site, not mvn..

But for sure i still am really confused how he knows that grouping at that time.. where does he get that info from?

[DEBUG] =======================================================================
[DEBUG] Resolving artifact com.servoy:javax.websocket:pom:1.1.0 from [central (https://repo.maven.apache.org/maven2, default, releases)]
[DEBUG] Using transporter WagonTransporter with priority -1.0 for https://repo.maven.apache.org/maven2
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for https://repo.maven.apache.org/maven2
Downloading from central: https://repo.maven.apache.org/maven2/com/servoy/javax.websocket/1.1.0/javax.websocket-1.1.0.pom
[DEBUG] Writing tracking file C:\Users\jcomp\.m2\repository\com\servoy\javax.websocket\1.1.0\javax.websocket-1.1.0.pom.lastUpdated
[WARNING] The POM for com.servoy:javax.websocket:jar:1.1.0 is missing, no dependency information available
[DEBUG] Dependency collection stats {ConflictMarker.analyzeTime=211000, ConflictMarker.markTime=122800, ConflictMarker.nodeCount=45, ConflictIdSorter.graphTime=71700, ConflictIdSorter.topsortTime=36800, ConflictIdSorter.conflictIdCount=30, ConflictIdSorter.conflictIdCycleCount=0, ConflictResolver.totalTime=463800, ConflictResolver.conflictItemCount=39, DefaultDependencyCollector.collectTime=337482900, DefaultDependencyCollector.transformTime=976900}
[DEBUG] org.sablo:sablo:eclipse-plugin:2022.6.0.3780_rc
[DEBUG]    junit:junit:jar:4.13.1:test
[DEBUG]       org.hamcrest:hamcrest-core:jar:1.3:test
[DEBUG]    org.mockito:mockito-core:jar:3.5.13:test
[DEBUG]       net.bytebuddy:byte-buddy:jar:1.10.15:test
[DEBUG]       net.bytebuddy:byte-buddy-agent:jar:1.10.15:test
[DEBUG]       org.objenesis:objenesis:jar:3.1:test
[DEBUG]    org.json:json:jar:20200518:test
[DEBUG]    org.skyscreamer:jsonassert:jar:1.5.0:test
[DEBUG]    p2.eclipse.plugin:jabsorb:jar:1.3.2.s6:system
[DEBUG]    p2.eclipse.plugin:jakarta.servlet-api:jar:4.0.0:system
[DEBUG]    p2.eclipse.plugin:slf4j.api:jar:1.8.0.beta3-SNAPSHOT:system
[DEBUG]    org.apache.aries.spifly:org.apache.aries.spifly.dynamic.bundle:jar:1.3.4:compile
[DEBUG]       org.ow2.asm:asm:jar:9.2:compile
[DEBUG]       org.ow2.asm:asm-util:jar:9.2:compile
[DEBUG]    p2.eclipse.plugin:org.objectweb.asm:jar:9.2.0.v20210813-1119:system
[DEBUG]    org.ow2.asm:asm-commons:jar:9.2:compile
[DEBUG]       org.ow2.asm:asm-tree:jar:9.2:compile
[DEBUG]       org.ow2.asm:asm-analysis:jar:9.2:compile
[DEBUG]    p2.eclipse.plugin:org.objectweb.asm.util:jar:9.2.0.v20210813-1119:system
[DEBUG]    p2.eclipse.plugin:org.eclipse.osgi:jar:3.17.200.v20220215-2237:system
[DEBUG]    p2.eclipse.plugin:org.apache.commons.io:jar:2.8.0.v20210415-0900:system
[DEBUG]    org.apache.logging.log4j:log4j-api:jar:2.17.2:compile
[DEBUG]    org.apache.logging.log4j:log4j-slf4j18-impl:jar:2.17.2:compile
[DEBUG]       org.slf4j:slf4j-api:jar:1.8.0-beta4:compile
[DEBUG]       org.apache.logging.log4j:log4j-core:jar:2.17.2:runtime
[DEBUG]    p2.eclipse.plugin:org.junit:jar:4.13.2.v20211018-1956:system
[DEBUG]    p2.eclipse.plugin:org.hamcrest.core:jar:1.3.0.v20180420-1519:system
[DEBUG]    p2.eclipse.plugin:org.objectweb.asm.analysis:jar:9.2.0.v20210813-1119:system
[DEBUG]    p2.eclipse.plugin:org.objectweb.asm.tree:jar:9.2.0.v20210813-1119:system
[DEBUG]    com.servoy:javax.websocket:jar:1.1.0:compile
[DEBUG] Resolving artifact com.servoy:javax.websocket:jar:1.1.0 from [central (https://repo.maven.apache.org/maven2, default, releases)]
[DEBUG] Using transporter WagonTransporter with priority -1.0 for https://repo.maven.apache.org/maven2
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for https://repo.maven.apache.org/maven2
Downloading from central: https://repo.maven.apache.org/maven2/com/servoy/javax.websocket/1.1.0/javax.websocket-1.1.0.jar
[DEBUG] Writing tracking file C:\Users\jcomp\.m2\repository\com\servoy\javax.websocket\1.1.0\javax.websocket-1.1.0.jar.lastUpdated
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  36.572 s
[INFO] Finished at: 2022-04-19T13:37:30+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project sablo: Could not resolve dependencies for project org.sablo:sablo:eclipse-plugin:2022.6.0.3780_rc: Could not find artifact com.servoy:javax.websocket:jar:1.1.0 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project sablo: Could not resolve dependencies for project org.sablo:sablo:eclipse-plugin:2022.6.0.3780_rc: Could not find artifact com.servoy:javax.websocket:jar:1.1.0 in central (https://repo.maven.apache.org/maven2)

**Comment by mickaelistria** [comment]
> But for sure i still am really confused how he knows that grouping at that time.. where does he get that info from?

From the `maven.groupId` p2 property for this artifact in content.xml.

**Reply by jcompagner** [reply]
ah. there it is
But does that mean that suddenly tycho wants to download it from mvn repo?
i don't want that, p2 site is the one it should get (thats also all signed)

**Reply by mickaelistria** [reply]
If there is a Maven artifact for this same GAV, then they're equivalent (it's a Maven rule) and Tycho should be able to either download one or the other. However, if the artifact doesn't exist on Maven repo, it shouldn't fail. So it could be that the issue you see is a bug in Tycho not properly handling the case of GAV being set while artifact not being present in the repo. If you can reduce it to the bare minimal reproducer, that would help.

**Reply by jcompagner** [reply]
i can try but first i had to go back to tycho 2.6, because it seems 2.7.x really doesn't work for me (this problem and the other one i mentioned in general/discussions)

But that was not something that just worked.. going back:

Caused by: java.lang.IllegalArgumentException: bundleLocation not found: C:\Users\jcomp\.m2\repository\org\eclipse\orbit\bundles\org.objectweb.asm\9.2.0-SNAPSHOT\org.objectweb.asm-9.2.0-SNAPSHOT.jar
    at org.eclipse.tycho.core.osgitools.EquinoxResolver.loadManifest (EquinoxResolver.java:497)
    at org.eclipse.tycho.core.osgitools.EquinoxResolver.newState (EquinoxResolver.java:352)
    at org.eclipse.tycho.core.osgitools.EquinoxResolver.newResolvedState (EquinoxResolver.java:126)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getResolverState (OsgiBundleProject.java:326)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:196)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:173)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.lambda$resolveProjects$0 (TychoMavenLifecycleParticipant.java:147)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant$$Lambda$205/00000000D66BB740.accept (Unknown Source)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:183)
    at java.util.stream.WhileOps$1$1.accept (WhileOps.java:99)
    at java.util.ArrayList$ArrayListSpliterator.tryAdvance (ArrayList.java:1632)
    at java.util.stream.ReferencePipeline.forEachWithCancel (ReferencePipeline.java:127)
    at java.util.stream.AbstractPipeline.copyIntoWithCancel (AbstractPipeline.java:515)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:501)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:487)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:150)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:173)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:239)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:497)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.resolveProjects (TychoMavenLifecycleParticipant.java:178)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:105)**

i had this failure even in an other branch...
That would be also horrible. if i upgrade tycho in our master branch, that downloads stuff and adjust the mvn repo
and then older branches don't build anymore....

i really had to delete prety much everything i could find in the maven repo for tych/p2/eclipse

then it started to work again..

**Reply by laeubi** [reply]
Please read https://github.com/eclipse/tycho/blob/master/RELEASE_NOTES.md#caution-when-switching-between-tycho-versions

**Reply by jcompagner** [reply]
ok so that means for me that i just can't update..
because we have 4 branches, 1 is master the other 3 are all current release or 2 LTS release branches 
Those will not be updated to the latest tycho, they will also be using older eclipses (those also won't update)

so this means on my jenkinks or locally we constantly need to delete that, but builds can be concurrently...

so i guess i need to keep sticking to 2.6 for quite a while :(

**Reply by laeubi** [reply]
There is nothing forcing you to upgrade. Beside that not all branches need to use the same (shared) m2 repository.

**Reply by jcompagner** [reply]
for use that means so many changes if we don;t use the same m2 repo
We have developers that do build (or partial builds) locally 
we have jenkins that builds all of 4 of them 
so having 4 m2 repo's would mean an huge amout of extra diskpace of a lot of duplicate stuff..

But yes it seems i need to stick to 2.6 for quite some time because of the 3 issues i now have

1> https://github.com/eclipse/tycho/discussions/894
2> the issue described here about tycho suddenly wanting to treath a p2 artifact as a mvn artifact (which it isn't really for us)
3> Tycho 2.7.x breaks the mvn repo for older tycho builds that we have..

i really hope that 2.6 will work for quite some time also on newer eclipses...

**Reply by laeubi** [reply]
Tycho has nothing to do with "newer eclipses" ...

**Comment by laeubi** [comment]
> Orignally that javax.websocket is created yes under that group. To generate our p2 site for 3rd party libs
> so it is there:

You generally do not need anymore generate "custom" p2 sites for 3rd party dependencies from maven. You can simply reference them directly in your target and then they won't get downloaded twice.

If you don't want to do that, simply add the developer repo so tycho can download from there:
```
<repositories>
        <repository>
            <id>developer_repo</id>
            <name>developer_repo</name>
            <url>https://developer.servoy.com/mvn_repository/</url>
        </repository>
```

Beside the duplicate download, tycho will still use the p2 artifact in all cases but this might change in 3.x see https://github.com/eclipse/tycho/discussions/818

**Reply by jcompagner** [reply]
but that is a problem
mvn artifacts are not signed and we need that.

So i need that tycho doesn't use the p2 site just as an index and then doesn't use the actual jars of that site at all...
but it should just use the p2 site if i gave him that completely, so the index and the actual jars that are given by that p2 site..

Now it seems it just uses p2 as an index... and just skips over the rest..

**Reply by laeubi** [reply]
Tycho uses the artifacts of the p2 repository. Beside that if you need signing in the product or site it should be performed there and not in an intermediate update-site.

**Reply by jcompagner** [reply]
no it does not, this is the exception i get:

[DEBUG] Resolving artifact com.servoy:javax.websocket:jar:1.1.0 from [central (https://repo.maven.apache.org/maven2, default, releases)]
[DEBUG] Using transporter WagonTransporter with priority -1.0 for https://repo.maven.apache.org/maven2
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for https://repo.maven.apache.org/maven2
Downloading from central: https://repo.maven.apache.org/maven2/com/servoy/javax.websocket/1.1.0/javax.websocket-1.1.0.jar
[DEBUG] Writing tracking file C:\Users\jcomp.m2\repository\com\servoy\javax.websocket\1.1.0\javax.websocket-1.1.0.jar.lastUpdated
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 36.572 s
[INFO] Finished at: 2022-04-19T13:37:30+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project sablo: Could not resolve dependencies for project org.sablo:sablo:eclipse-plugin:2022.6.0.3780_rc: Could not find artifact com.servoy:javax.websocket:jar:1.1.0 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project sablo: Could not resolve dependencies for project org.sablo:sablo:eclipse-plugin:2022.6.0.3780_rc: Could not find artifact com.servoy:javax.websocket:jar:1.1.0 in central (https://repo.maven.apache.org/maven2)


that com.servoy:javax.websocket:jar

is just on the our p2 site:

https://developer.servoy.com/p2_repository/ship_plugins/2022.06/plugins/javax.websocket_1.1.0.jar

but for some reason it wants to resolve it through a mvn repo and yes it is not on there an then the build fails.

**Reply by jcompagner** [reply]
besides that how do you sign a product and then only the once that needs signing?
(i don't want to resign on every build every jar, including the once that are already signed and just fine)

i rather do the signing once, so the source, not everytime i make a installer from those artifacts..

**Reply by laeubi** [reply]
> no it does not, this is the exception i get:

The exception does not tell what you claim, this is a maven error and not a tycho error because you are missing the repository where this artifact originated from.

**Reply by laeubi** [reply]
> besides that how do you sign a product and then only the once that needs signing?

See for example https://github.com/eclipse-m2e/m2e-core/blob/25d6833f9cf78e0c423155c9d8381f0db8eaa9d0/org.eclipse.m2e.repository/pom.xml#L38

**Reply by jcompagner** [reply]
> 

right, and tycho or mvn shouldn't do anything with that
That is just locally generated, don't use anything with that mvn group or whatever
just get the jar file from the p2 site just like Tycho 2.6 does...

can i get that for example "maven.groupId" is NOT set in the content.xml of the p2 site?
because then at least tycho has no idea that this is a mvn thing, and thats what i want. I just want tycho to see this as a P2 artifact, don't assume anything else..

**Reply by jcompagner** [reply]
> > besides that how do you sign a product and then only the once that needs signing?
> 
> See for example https://github.com/eclipse-m2e/m2e-core/blob/25d6833f9cf78e0c423155c9d8381f0db8eaa9d0/org.eclipse.m2e.repository/pom.xml#L38

hmm that is org.eclipse.cbi.maven.plugins
not sure if i can use that...
isn't that the eclipse build infrastructure ?
what are those:

```
<processMainArtifact>false</processMainArtifact>
							<processAttachedArtifacts>false</processAttachedArtifacts>
```

2 properties mean?
but as i said i don't like that
its just better to make sure the src is signed and not needing to do this in every build (builds for use are constantly created at least every day)

**Reply by laeubi** [reply]
> right, and tycho or mvn shouldn't do anything with that
> That is just locally generated, don't use anything with that mvn group or whatever

That's obviously not true, you depend on maven dependencies and those are not local to the build. Local items are not added, only those that are consumed from an external site.

And as mentioned before they are *downloaded* but not *used* still maven wan't to download them, so if you simply add the repository it will be download to disk but tycho will use the p2 items.

**Reply by jcompagner** [reply]
no it is locally
i have a local m2 repo
where i put some jars (mostly forked jars, github projects) 
from that i generate the p2 site..

yes i can then make a mvn site in the cloud to really also have all those, but until now that is not really needed.

so the mvn site is just some local intermediate thing, has no meaning or function after i created the p2 site from it.

**Reply by laeubi** [reply]
Here you reference a maven dependency that seems not be local to the build:
https://github.com/Servoy/servoy-eclipse/blob/master/shipplugins/category.xml#L32
because otherwise tycho won't add this as a regular maven dependency.
